## Logistic Regression

In this notebook I am going to work through an example of classifier using <a href="https://en.wikipedia.org/wiki/Logistic_regression">logistic regression</a>. The focus will be on the practical application rather that working through the mathematics and theory behind it. The plan for this exercise is to use the scikit learn module from Python in conjunction with the <a href="https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+(Diagnostic)">Wisconsin breast cancer dataset</a> to create, train, test and evaluate a classifier.

<img src="img/logReg.jpg">

I will start by importing the required libraries for the project.

### Import

In [1]:
# import the required libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.rc("font", size=14)
import seaborn as sns
sns.set(style="white")
sns.set(style="whitegrid", color_codes=True)
# import required sklearn modules
from sklearn import preprocessing
from sklearn.linear_model import LogisticRegression
import sklearn.model_selection as mod
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

### Dataset

For this example I'm going to use the breast cancer dataset

In [6]:
# load the dataset using the sklearn package - https://scikit-learn.org/stable/datasets/index.html
from sklearn.datasets import load_breast_cancer
breast_cancer = load_breast_cancer()

# create a dataframe with the data from the dataset
df = pd.DataFrame(breast_cancer.data, columns=breast_cancer.feature_names)
tar = pd.DataFrame(breast_cancer.target)
# add the median value to the dataframe
df["Diagnosis"] = breast_cancer.target

# show the head of the dataset
df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  Diagnosis  
0          0.4601                  0.11890          0  
1          0.2750                  0.08902          0  
2          0.3613                  0.08758          0  
3          0.6638                  0.17300          0  
4          0.2364                  0.07678          0  

[5 rows x 31 columns]

In [3]:
# print the shape of the dataset
df.shape

(569, 31)

### Summary of the Dataset
* The dataset contains 569 data points, and 31 columns of data for each data point. 
* Each data point represents a patients and their tumour.
* The data features contain various measurements for each tumour, and the target contains the diagnosis for each patient. 
* If the value for the Diagnosis was '0' then the tumour was malignant, if it was '1' then the tumour was benign.

### Fit the model
Next we will use sklearn to fit a linear regression model. The steps for this are as follows:
1. Define the inputs and the outputs for the model.
1. Split the dataset into training data and testing data (I will use an 80/20 split for this)
1. Fit a model using the training data.
1. Use the model, and to predict tumour classifications for the test data.
1. Compare these to the actual result to see how accurate the model is.

In [4]:
# define the inputs and outputs
outputs = df.iloc[:,-1]
inputs = df.iloc[:,:-1]

# split the data into testing and training data
inputs_train, inputs_test, outputs_train, outputs_test = mod.train_test_split(inputs, outputs, test_size=0.2)

# fit a logistic regression model using the training data
logReg = LogisticRegression().fit(inputs_train, outputs_train)

# make a prediction
test_predict = logReg.predict(inputs_test)      

# compare the predictions with the actual result
score =accuracy_score(outputs_test,test_predict)
print("The accuracy of this logisitic regression model is: ", round(score,2))

The accuracy of this logisitic regression model is:  0.95


/Users/patrickmoore/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


This model has correctly predicted whether or not a tumour is malignant about 95% of the time, next I will consider the confusion matrix to see how the errors in the model are distributed.

In [5]:
confusion_matrix(outputs_test, test_predict)

array([[38,  4],
       [ 2, 70]])


### Conclusion

### End